File: content/snippets/cpp/interview_lcci/16/20_t9.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

char mapDigit(char ch) {
    if (ch <= 'c') return '2';
    if (ch <= 'f') return '3';
    if (ch <= 'i') return '4';
    if (ch <= 'l') return '5';
    if (ch <= 'o') return '6';
    if (ch <= 's') return '7';
    if (ch <= 'v') return '8';
    return '9';
}

vector<string> getValidT9Words(const string& num, const vector<string>& words) {
    vector<string> ans;
    for (const string& word : words) {
        if (word.size() != num.size()) continue;
        bool ok = true;
        for (int i = 0; i < static_cast<int>(word.size()); ++i) {
            if (mapDigit(word[i]) != num[i]) {
                ok = false;
                break;
            }
        }
        if (ok) ans.push_back(word);
    }
    return ans;
}

int main() {
    vector<string> ans = getValidT9Words("8733", {"tree", "used", "true"});
    cout << "[";
    for (int i = 0; i < static_cast<int>(ans.size()); ++i) {
        if (i) cout << ", ";
        cout << ans[i];
    }
    cout << "]" << endl;
    return 0;
}

File: content/snippets/scala3/interview_lcci/16/20_t9.scala
object Lcci1620T9:
    private def mapDigit(ch: Char): Char = {
        if (ch <= 'c') '2'
        else if (ch <= 'f') '3'
        else if (ch <= 'i') '4'
        else if (ch <= 'l') '5'
        else if (ch <= 'o') '6'
        else if (ch <= 's') '7'
        else if (ch <= 'v') '8'
        else '9'
    }

    def getValidT9Words(num: String, words: Array[String]): Array[String] =
        words.filter { word =>
            word.length == num.length && word.indices.forall(i => mapDigit(word(i)) == num(i))
        }

    @main def main(): Unit =
        println(getValidT9Words("8733", Array("tree", "used", "true")).mkString("[", ", ", "]"))
